File: scra.php
<?php
function match($desk, $word) {
    if($desk == "") return false;
    if($word == "") return true;

    $w = substr($word, 0, 1);
    $ws = substr($word, 1);

    $pos = strpos($desk, $w);
    if($pos !== false) {
        $desk = substr($desk, 0, $pos).substr($desk, $pos + 1);
        return match($desk, $ws);
    }

    $pos = strpos($desk, '*');
    if($pos !== false) {
        $desk = substr($desk, 0, $pos).substr($desk, $pos + 1);
        return match($desk, $ws);
    }

    return false;
}

function search($tokens, $dictionary) {
    $filter = function($word) use($tokens) { return match($tokens, $word); };

    return array_filter($dictionary, $filter);
}

function prettyPrint($words) {
    $compareLength = function($a, $b) { return strlen($a) - strlen($b); };

    usort($words, $compareLength);
    print(implode("\n", $words)."\n");
}

if(isset($argv[1])) {
    $firstArg = $argv[1];
} else {
    exit("One parameter needed");
}

$dictionary = explode("\n", file_get_contents("ODS6.txt"));

$results = search(strtoupper($firstArg), $dictionary);

prettyPrint($results);

File: scra.py
#!/usr/bin/env python

from sys import argv, exit

def match(desk, word):
    if desk == "": return False
    if word == "": return True

    w = word[0]
    ws = word[1:]
    if w in desk:
        return match(desk.replace(w, '', 1), ws)

    if '*' in desk:
        return match(desk.replace('*', '', 1), ws)

    return False

def search(tokens, dictionary):
    return [word for word in dictionary if match(tokens, word)]

def prettyPrint(words):
    print "\n".join(sorted(words, key=len))

if __name__ == '__main__':
    if len(argv) == 2:
        firstArg = argv[1]
    else:
        exit("One parameter needed")

    dictionary = open("ODS6.txt", "r").read().splitlines()

    results = search(firstArg.upper(), dictionary)

    prettyPrint(results)

